lib/AST/ASTContext.cpp

```diff
@@ -8407,6 +8413,48 @@ bool ASTContext::isEqualIgnoringChecked(QualType T1, QualType T2) const {
   }
 }
 
+// For the Checked C extension, compute whether a type is allowed to be an
+// argument or return type for a no-prototype function.   This computes the
+// set of allowed types described in Section 5.5 of the Checked C
+// specification.
+bool ASTContext::isNotAllowedForNoProtoTypeFunction(QualType QT) const {
+  if (const PointerType *PT = QT->getAs<PointerType>()) {
+    if (PT->isChecked())
+      return true;
+    if (PT->isFunctionPointerType())
+      return isNotAllowedForNoProtoTypeFunction(PT->getPointeeType());
+    // Unchecked pointer types are allowed
+    return false;
+  } else if (QT->isCheckedArrayType())
```

Comment:
Can you have an unchecked array of function pointers, where the function pointers have params or return types that are checked?

I get nervous seeing this sort of logic so specific to a scenario (dealing with no-prototype functions).  It might be simpler and easier to maintain to have a single helper method somewhere in the system that, once and for all, determines if a given type--in any way--contains any sort of checked types or bounds expressions.  And perhaps we can implement that using a proper AST walker?

And feel free to tell me my anxiety about this is unnecessary right now.